### PR TITLE
DataGrid: Implement insert changes assignment

### PR DESCRIPTION
**Pull request by jtoming830**

### CODE REVIEW

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 1]
```diff
@@ -816,11 +814,12 @@ const EditingController = modules.ViewController.inherit((function() {
...
             return item;
         },
-        _getLoadedRowIndexByInsertInfo: function(items, change, key) {
+        _getLoadedRowIndex: function(items, change, key) {
             const dataController = this._dataController;
             const loadedRowIndexOffset = dataController.getRowIndexOffset(true);
-            const insertInfo = this._getInternalData(key)?.insertInfo;
-            let loadedRowIndex = (insertInfo ? insertInfo.dataRowIndex : 0) - loadedRowIndexOffset;
+            const changes = this.getChanges();
+            const index = change ? changes.filter(editChange => editChange.key === key)[0].index : 0;
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: key may be complex

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by NickMitrokhin — DISMISSED** [review]

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 2]
```diff
@@ -921,47 +916,61 @@ const EditingController = modules.ViewController.inherit((function() {
...
+            const row = rows[rowIndex];
+
+            if(row && (!row.isEditing && row.rowType === 'detail' || row.rowType === 'detailAdaptive')) {
+                rowIndex++;
+            }
+
+            return rowIndex;
+        },
+
+        _addInsertInfo: function(change = {}) {
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: Why is default value {} is here if change is always passed to _addInsertInfo?

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 3]
```diff
@@ -921,47 +916,61 @@ const EditingController = modules.ViewController.inherit((function() {
...
-            }
+            const insertInfo = {
+                parentKey: change.parentKey
+            };
 
-            insertInfo.dataRowIndex = dataController.getRowIndexOffset() + rows.filter(function(row, index) {
-                return index < insertInfo.rowIndex && (row.rowType === 'data' && !row.isNewRow || row.rowType === 'group');
-            }).length;
+            change.index = change.index ?? this._calculateIndex(rowIndex);
+            change.pageIndex = change.pageIndex ?? dataController.pageIndex();
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js]: Move change properties modification (index, pageIndex) from _createInsertInfo method to _addInsertInfo. The _createInsertInfo method will be cleaner.

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 4]
```diff
@@ -921,47 +916,61 @@ const EditingController = modules.ViewController.inherit((function() {
...
+            return rowIndex;
+        },
+
+        _addInsertInfo: function(change = {}) {
             let insertInfo;
+            let rowIndex;
+            let { key } = change;
 
             if(!isDefined(key)) {
                 key = String(new Guid());
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js:956]: Why key is not added to change here but key adds to change after each _addInsertInfo call?

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by NickMitrokhin — APPROVED** [review]

File: js/ui/grid_core/ui.grid_core.editing.js [hunk 5]
```diff
@@ -953,19 +945,30 @@ const EditingController = modules.ViewController.inherit((function() {
...
 
             insertInfo = this._getInternalData(key)?.insertInfo;
             if(!isDefined(insertInfo)) {
                 rowIndex = this._getCorrectedInsertRowIndex(change.parentKey);
                 insertInfo = this._createInsertInfo(rowIndex, change);
+
+                change.index = change.index ?? this._calculateIndex(rowIndex);
+
+                if(this.option('scrolling.mode') === 'virtual') {
+                    change.pageIndex = Math.floor(change.index / dataController.pageSize());
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.editing.js:968]: What is about infinite scrolling? It is very similar to virtual during paging.

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by vconst — APPROVED** [review]

**Pull request merged**